src/nlsolve/anderson.jl

```diff
@@ -42,15 +48,28 @@ function (S::NLAnderson{false,<:NLSolverCache})(integrator)
   fail_convergence = false
   while (do_anderson || iter < min_iter) && iter < max_iter
     iter += 1
-    u = tmp + γ*z
-    z₊ = dt*f(u, p, tstep)
+    u = @. tmp + γ*z
+    if mass_matrix == I
+      z₊ = dt .* f(u, p, tstep)
+    else
+      mz = mass_matrix * z
+      z₊ = dt .* f(u, p, tstep) .- mz .+ z
+    end
     gs[1] = z₊
 
     mk = min(S.n, iter-1)
-    residuals[1:mk] .= (gs[2:mk+1] .- zs[2:mk+1]) .- (gs[1] - zs[1])
-    alphas[1:mk] .= residuals[1:mk] \ [(zs[1] - gs[1])]
+    ztmp = zs[1] .- gs[1]
+    for i in 2:mk+1
+      @. residuals[:,i-1] = gs[i] - zs[i] + ztmp
+    end
+    resqr = qr(residuals[:,1:mk], Val(true))
```

Comment:
Using `qr` won't help. You have `residuals = fill(zero(eltype(z)), length(z), S.n+1)` anyway.

Reply:
Using `qr` instead of `qr!` here does nothing other than increases memory allocation. However, let's leave it as-is, and write our custom QR factorization in another PR.